File: gtf_to_duckdb.cpp
#include <duckdb.hpp>               // DuckDB C++ API header
#include <ray/api.h>                // Experimental Ray C++ API header
#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <fstream>
#include <iostream>
#include <mutex>
#include <queue>
#include <sstream>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

// -----------------------------------------------------------------------------
// Thread-safe queue
template<typename T>
class TSQueue {
public:
    void push(const T &item) {
        std::unique_lock<std::mutex> lock(mutex_);
        queue_.push(item);
        cond_.notify_one();
    }
    
    bool try_pop(T &item) {
        std::unique_lock<std::mutex> lock(mutex_);
        if (queue_.empty())
            return false;
        item = queue_.front();
        queue_.pop();
        return true;
    }
    
    bool empty() {
        std::unique_lock<std::mutex> lock(mutex_);
        return queue_.empty();
    }
    
private:
    std::queue<T> queue_;
    std::mutex mutex_;
    std::condition_variable cond_;
};

// -----------------------------------------------------------------------------
// Process a single line from the GTF file and return an SQL INSERT statement.
std::string process_line(const std::string &line, const std::string &tablename) {
    if (line.empty() || line[0] == '#')
        return "";
    
    // Remove trailing newline (if any) and tokenize by tab.
    std::istringstream iss(line);
    std::vector<std::string> fields;
    std::string token;
    while (std::getline(iss, token, '\t')) {
        fields.push_back(token);
    }
    if (fields.size() < 9)
        throw std::runtime_error("Error: Line has less than 9 tab-separated fields.");
    
    // Process first eight fields.
    std::vector<std::pair<std::string, std::string>> data;
    std::vector<std::string> keys_list = {"Chromosome", "Source", "Feature", "Start", "End", "Score", "Strand", "Frame"};
    for (size_t i = 0; i < keys_list.size(); i++) {
        data.push_back({keys_list[i], fields[i]});
    }
    
    // Adjust the Start field (subtract 1)
    try {
        int start_val = std::stoi(data[3].second);
        data[3].second = std::to_string(start_val - 1);
    } catch (...) {
        throw std::runtime_error("Error processing 'Start' value: " + data[3].second);
    }
    
    // Process attribute field (9th field)
    std::string attributes = fields[8];
    std::istringstream attr_stream(attributes);
    std::string attribute;
    while (std::getline(attr_stream, attribute, ';')) {
        // Trim leading white-space:
        size_t first = attribute.find_first_not_of(" \t\n\r");
        if (first == std::string::npos)
            continue;
        attribute = attribute.substr(first);
        // Split on first space
        size_t pos = attribute.find(' ');
        if (pos != std::string::npos) {
            std::string key = attribute.substr(0, pos);
            std::string value = attribute.substr(pos + 1);
            // Remove any quotation marks
            value.erase(std::remove(value.begin(), value.end(), '"'), value.end());
            // Trim spaces from value.
            size_t val_start = value.find_first_not_of(" \t\n\r");
            size_t val_end = value.find_last_not_of(" \t\n\r");
            if (val_start != std::string::npos && val_end != std::string::npos)
                value = value.substr(val_start, val_end - val_start + 1);
            data.push_back({key, value});
        }
    }
    
    // Build the INSERT statement. Column names are double-quoted to avoid conflicts with reserved words.
    std::ostringstream oss;
    oss << "INSERT INTO " << tablename << " (";
    bool first_col = true;
    for (auto &p : data) {
        if (!first_col)
            oss << ", ";
        first_col = false;
        oss << "\"" << p.first << "\"";
    }
    oss << ") VALUES (";
    bool first_val = true;
    for (auto &p : data) {
        if (!first_val)
            oss << ", ";
        first_val = false;
        // Escape single quotes: replace each ' with ''
        std::string val = p.second;
        size_t pos = 0;
        while ((pos = val.find("'", pos)) != std::string::npos) {
            val.insert(pos, "'");
            pos += 2;
        }
        oss << "'" << val << "'";
    }
    oss << ");";
    return oss.str();
}

// -----------------------------------------------------------------------------
// Define a Ray remote function to process a batch of lines.
// The function takes a vector of lines and the table name, and returns a single string
// with the concatenated SQL INSERT statements.
std::string ProcessBatch(const std::vector<std::string> &lines_batch) {
    std::string table_name = "human"; // Default table name
    std::ostringstream batch_sql;
    for (const auto &line : lines_batch) {
        std::string stmt = process_line(line, table_name);
        if (!stmt.empty())
            batch_sql << stmt << "\n";
    }
    return batch_sql.str();
}
RAY_REMOTE(ProcessBatch);

// -----------------------------------------------------------------------------
// Main function
int main(int argc, char **argv) {
    if (argc != 4) {
        std::cerr << "Usage: " << argv[0] << " <db_name> <table_name> <input_file>" << std::endl;
        return 1;
    }
    std::string db_name = argv[1];
    std::string table_name = argv[2];
    std::string input_file = argv[3];

    // Initialize Ray.
    ray::Init();

    auto start = std::chrono::high_resolution_clock::now();

    // Connect to DuckDB.
    // For a file-based database, pass the filename; otherwise, use nullptr for an in-memory DB.
    duckdb::DuckDB db(db_name.c_str());
    duckdb::Connection conn(db);

    // Drop existing table.
    conn.Query("DROP TABLE IF EXISTS " + table_name + ";");

    // Create table.
    std::string create_query;
    if (table_name == "human") {
        create_query = "CREATE TABLE IF NOT EXISTS " + table_name + " ("
                       "\"Chromosome\" TEXT, "
                       "\"Source\" TEXT, "
                       "\"Feature\" TEXT, "
                       "\"Start\" INTEGER, "
                       "\"End\" INTEGER, "
                       "\"Score\" TEXT, "
                       "\"Strand\" TEXT, "
                       "\"Frame\" TEXT, "
                       "\"gene_id\" TEXT, "
                       "\"gene_version\" TEXT, "
                       "\"gene_source\" TEXT, "
                       "\"gene_biotype\" TEXT, "
                       "\"transcript_id\" TEXT, "
                       "\"transcript_version\" TEXT, "
                       "\"transcript_source\" TEXT, "
                       "\"transcript_biotype\" TEXT, "
                       "\"tag\" TEXT, "
                       "\"transcript_support_level\" TEXT, "
                       "\"exon_number\" TEXT, "
                       "\"exon_id\" TEXT, "
                       "\"exon_version\" TEXT, "
                       "\"gene_name\" TEXT, "
                       "\"transcript_name\" TEXT, "
                       "\"protein_id\" TEXT, "
                       "\"protein_version\" TEXT, "
                       "\"ccds_id\" TEXT"
                       ");";
    } else if (table_name == "mouse") {
        create_query = "CREATE TABLE IF NOT EXISTS " + table_name + " ("
                       "\"Chromosome\" TEXT, "
                       "\"Source\" TEXT, "
                       "\"Feature\" TEXT, "
                       "\"Start\" INTEGER, "
                       "\"End\" INTEGER, "
                       "\"Score\" TEXT, "
                       "\"Strand\" TEXT, "
                       "\"Frame\" TEXT, "
                       "\"gene_id\" TEXT, "
                       "\"gene_type\" TEXT, "
                       "\"gene_name\" TEXT, "
                       "\"level\" TEXT, "
                       "\"mgi_id\" TEXT, "
                       "\"havana_gene\" TEXT, "
                       "\"transcript_id\" TEXT, "
                       "\"transcript_type\" TEXT, "
                       "\"transcript_name\" TEXT, "
                       "\"transcript_support_level\" TEXT, "
                       "\"tag\" TEXT, "
                       "\"havana_transcript\" TEXT, "
                       "\"exon_number\" TEXT, "
                       "\"exon_id\" TEXT, "
                       "\"protein_id\" TEXT, "
                       "\"ccdsid\" TEXT, "
                       "\"ont\" TEXT"
                       ");";
    } else {
        std::cerr << "Unsupported table name: " << table_name << std::endl;
        return 1;
    }
    conn.Query(create_query);
    
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - start);
    std::cout << "Time taken to create table: " << elapsed.count() << "ms" << std::endl;

    // Set up a thread-safe queue for SQL batches and a flag for read completion.
    TSQueue<std::string> sql_queue;
    std::atomic<bool> read_done(false);

    // Writer thread function: pop SQL batches from the queue and execute each statement.
    auto writer_thread_func = [&]() {
        while (!read_done.load() || !sql_queue.empty()) {
            std::string batch_sql;
            if (sql_queue.try_pop(batch_sql)) {
                auto start_time = std::chrono::high_resolution_clock::now();
                std::istringstream iss(batch_sql);
                std::string stmt;
                while (std::getline(iss, stmt)) {
                    if (stmt.find_first_not_of(" \t") != std::string::npos) {
                        conn.Query(stmt);
                    }
                }
                auto commit_time = std::chrono::high_resolution_clock::now();
                auto diff = std::chrono::duration_cast<std::chrono::milliseconds>(commit_time - start_time);
                std::cout << "Pushed a batch to DuckDB in " << diff.count() << "ms" << std::endl;
            } else {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }
        }
    };

    // Reader thread function: read the input file, divide into batches, and process via Ray tasks.
    auto reader_thread_func = [&]() {
        std::ifstream infile(input_file);
        if (!infile) {
            std::cerr << "Failed to open input file: " << input_file << std::endl;
            return;
        }
        // Although the Python code uses a byte-based batch size,
        // here we simply read line-by-line and group into chunks.
        const size_t chunk_size = 8000; // lines per batch
        std::vector<std::string> all_lines;
        std::string line;
        while (std::getline(infile, line)) {
            all_lines.push_back(line);
            if (all_lines.size() >= chunk_size) {
                // Submit a Ray remote task for this batch.
                std::vector<ray::ObjectRef<std::string>> futures;
                futures.push_back(ray::Task(ProcessBatch).Remote(all_lines));
                std::string batch_sql;
                for (auto &future : futures) {
                    batch_sql += *future.Get();
                }
                sql_queue.push(batch_sql);
                all_lines.clear();
            }
        }
        // Process any remaining lines.
        if (!all_lines.empty()) {
            auto future = ray::Task(ProcessBatch).Remote(all_lines);
            std::string batch_sql = *future.Get();
            sql_queue.push(batch_sql);
        }
        read_done.store(true);
        std::cout << "Reader thread finished reading file" << std::endl;
    };

    // Start the reader and writer threads.
    std::thread reader_thread(reader_thread_func);
    std::thread writer_thread(writer_thread_func);
    reader_thread.join();
    writer_thread.join();

    // Create indices.
    start = std::chrono::high_resolution_clock::now();
    try {
        std::string index_query = "CREATE INDEX \"ix_" + table_name + "_Chromosome_Strand\" ON \"" + table_name + "\" (\"Chromosome\", \"Strand\");";
        conn.Query(index_query);
    } catch (std::exception &e) {
        std::cout << "Warning: Skipped index creation due to error: " << e.what() << std::endl;
    }
    elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - start);
    std::cout << "Time taken to create indices: " << elapsed.count() << "ms" << std::endl;

    // For file-based databases (db_name containing ".ddb"), the connection will be saved when closed.
    // DuckDB cleans up when objects go out of scope.
    
    // Shutdown Ray.
    ray::Shutdown();
    return 0;
}

File: gtf_to_duckdb.py
import duckdb
import sys
import time
import ray
import threading
import queue

def process_line(line, tablename):
    """
    Process a single GTF file line and return an SQL INSERT statement.
    
    The function splits the line on tab characters,
    processes the first 8 fields and subtracts 1 from the Start field.
    Then it parses the attribute field (the ninth column) by splitting on ';'
    and further splitting each attribute on the first space.
    It finally builds an INSERT statement inserting the collected key/value pairs.
    """
    # dont process comment lines i.e lines starting with '#'
    if line.startswith('#'):
        return None
    line = line.rstrip("\n")
    # Strip the trailing newline and split by tab:
    fields = line.strip().split('\t')
    if len(fields) < 9:
        raise ValueError("Error: Line has less than 9 tab-separated fields.")
    
    # Prepare the first columns in order:
    keys_list = ["Chromosome", "Source", "Feature", "Start", "End", "Score", "Strand", "Frame"]
    data = {}
    for i, key in enumerate(keys_list):
        data[key] = fields[i]
    
    # Adjust the Start field (subtract 1)
    try:
        data["Start"] = str(int(data["Start"]) - 1)
    except Exception as e:
        raise ValueError(f"Error processing 'Start' value: {data['Start']}") from e
    
    # Process the attribute field (9th field)
    attributes = fields[8]
    for attribute in attributes.split(';'):
        attribute = attribute.lstrip()  # Remove leading whitespace
        if not attribute:
            continue  # Skip empty tokens
        # Split attribute into key and value at first space
        parts = attribute.split(' ', 1)
        if len(parts) == 2:
            key, value = parts
            # Remove any quotation marks and extra whitespace from the value
            value = value.replace('"', '').strip()
            data[key] = value

    # Build the INSERT statement.
    # Note: This builds the SQL command as a string, similar to the C++ code.
    # In production code you would generally want to use parameterized queries.
    columns = ", ".join(f'"{k}"' for k in data.keys())
    # Escape single quotes in values by doubling them up:
    values = ", ".join("'" + v.replace("'", "''") + "'" for v in data.values())
    sql = f"INSERT INTO {tablename} ({columns}) VALUES ({values});"
    return sql

@ray.remote
def process_batch(lines_batch, table_name):
    # Process each line in the batch
    sql_statements = ""
    for line in lines_batch:
        # You can call your original process_line function here.
        # Or, directly implement the processing logic.
        sql_statement = process_line(line, table_name)
        if sql_statement:
            sql_statements += sql_statement + "\n"
    # Return the accumulated SQL statements for this batch
    return sql_statements

def run_gtf_to_sql(db_name, table_name, input_file, batch_size=100000000, chunk_size=8000):
    start_time = time.time()
    conn = duckdb.connect(db_name)
    cur = conn.cursor()

    # Drop the table if it exists
    drop_query = f"DROP TABLE IF EXISTS {table_name};"
    cur.execute(drop_query)

    # Create table
    if table_name == "human":
        create_query = f'''
        CREATE TABLE IF NOT EXISTS {table_name} (
            "Chromosome" TEXT,
            "Source" TEXT,
            "Feature" TEXT,
            "Start" INTEGER,
            "End" INTEGER,
            "Score" TEXT,
            "Strand" TEXT,
            "Frame" TEXT,
            "gene_id" TEXT,
            "gene_version" TEXT,
            "gene_source" TEXT,
            "gene_biotype" TEXT,
            "transcript_id" TEXT,
            "transcript_version" TEXT,
            "transcript_source" TEXT,
            "transcript_biotype" TEXT,
            "tag" TEXT,
            "transcript_support_level" TEXT,
            "exon_number" TEXT,
            "exon_id" TEXT,
            "exon_version" TEXT,
            "gene_name" TEXT,
            "transcript_name" TEXT,
            "protein_id" TEXT,
            "protein_version" TEXT,
            "ccds_id" TEXT
        );
        '''
    elif table_name == "mouse":
        create_query = f'''
        CREATE TABLE IF NOT EXISTS {table_name} (
            "Chromosome" TEXT,
            "Source" TEXT,
            "Feature" TEXT,
            "Start" INTEGER,
            "End" INTEGER,
            "Score" TEXT,
            "Strand" TEXT,
            "Frame" TEXT,
            "gene_id" TEXT,
            "gene_type" TEXT,
            "gene_name" TEXT,
            "level" TEXT,
            "mgi_id" TEXT,
            "havana_gene" TEXT,
            "transcript_id" TEXT,
            "transcript_type" TEXT,
            "transcript_name" TEXT,
            "transcript_support_level" TEXT,
            "tag" TEXT,
            "havana_transcript" TEXT,
            "exon_number" TEXT,
            "exon_id" TEXT,
            "protein_id" TEXT,
            "ccdsid" TEXT,
            "ont" TEXT
        );
        '''
    else:
        raise ValueError(f"Unsupported table name: {table_name}")
    cur.execute(create_query)

    conn.commit()
    elapsed = time.time() - start_time
    print(f"Time taken to create table: {elapsed*1000:.0f}ms")

    sql_queue = queue.Queue(maxsize=1)
    read_done = threading.Event()

    def writer_thread_func():
        while not (read_done.is_set() and sql_queue.empty()):
            try:
                batch_sql_statements = sql_queue.get(timeout=1)
            except queue.Empty:
                continue
            start_time = time.time()
            conn.begin()
            for stmt in batch_sql_statements.splitlines():
                if stmt.strip():
                    cur.execute(stmt)
            conn.commit()
            elapsed = time.time() - start_time
            print(f"Pushed a batch to DuckDB in {elapsed*1000:.0f}ms")

    def reader_thread_func():
        with open(input_file, "r") as f:
            while True:
                start_time = time.time()
                lines = f.readlines(batch_size)
                if not lines:
                    break
                elapsed = time.time() - start_time
                print(f"Read {len(lines)} lines from file in {elapsed*1000:.0f}ms")

                start_time = time.time()
                futures = []
                for i in range(0, len(lines), chunk_size):
                    lines_batch = lines[i:i+chunk_size]
                    futures.append(process_batch.remote(lines_batch, table_name))
                elapsed = time.time() - start_time
                print(f"Submitted {len(lines)} lines to ray in {elapsed*1000:.0f}ms")

                start_time = time.time()
                sql_statements = ray.get(futures)
                elapsed = time.time() - start_time
                print(f"Processed {len(lines)} lines in {elapsed*1000:.0f}ms")

                start_time = time.time()
                batch_sql_statements = "".join(sql_statements)
                elapsed = time.time() - start_time
                print(f"Flattened SQL statements in {elapsed*1000:.0f}ms")

                sql_queue.put(batch_sql_statements)
        read_done.set()
        print("Reader thread finished reading file")

    reader_thread = threading.Thread(target=reader_thread_func)
    writer_thread = threading.Thread(target=writer_thread_func)
    reader_thread.start()
    writer_thread.start()

    reader_thread.join()
    writer_thread.join()

    # Create indices
    start_time = time.time()
    try:
        index_query2 = f'CREATE INDEX "ix_{table_name}_Chromosome_Strand" ON "{table_name}" ("Chromosome", "Strand");'
        cur.execute(index_query2)
    except Exception as e:
        print(f"Warning: Skipped index creation due to error: {e}")
    conn.commit()
    elapsed = time.time() - start_time
    print(f"Time taken to create indices: {elapsed*1000:.0f}ms")

    # Close the connection only if the database is file-based.
    if db_name.endswith(".ddb"):
        conn.close()

def main():
    if len(sys.argv) != 4:
        print("Usage: python gtf_to_sql.py <db_name> <table_name> <input_file>")
        sys.exit(1)

    db_name = sys.argv[1]
    table_name = sys.argv[2]
    input_file = sys.argv[3]
    run_gtf_to_sql(db_name, table_name, input_file)

if __name__ == "__main__":
    start_time = time.time()
    ray.init()
    elapsed = time.time() - start_time
    print(f"Ray initialized in {elapsed*1000:.0f}ms")
    # Run the main function
    start_time = time.time()
    main()
    elapsed = time.time() - start_time
    print(f"Main function completed in {elapsed*1000:.0f}ms")
    # Shutdown Ray
    start_time = time.time()
    ray.shutdown()
    elapsed = time.time() - start_time
    print(f"Ray shutdown in {elapsed*1000:.0f}ms")
